# Recommended System Armut Co.

### Business Problem

Armut, Turkey's largest online service platform, brings together service providers and those who want to receive services. It provides easy access to services such as cleaning, renovation, transportation with a few taps on your computer or smartphone. We want to create a product recommendation system with Association Rule Learning using a data set that includes service users and the services and categories that these users have received.

### History of Dataset
The dataset consists of the services received by the customers and the categories of these services. It contains the date and time information of each service received.
### **Variables**
- **UserId**: Customer number
- **ServiceId**: Anonymized services belonging to each category (Example: Sofa washing service under the cleaning category)
- **A ServiceId** can be found under different categories and refers to different services under different categories.
- **(Example**: CategoryId 7 ServiceId 4 refers to honeycomb cleaning while CategoryId 2 ServiceId 4 refers to furniture assembly)
- **CategoryId**: Anonymized categories (Example: Cleaning, transportation, renovation category)
- **CreateDate**: Date the service was purchased


### Necessary Lib.

In [14]:
import pandas as pd
pd.set_option('display.max_columns', None)
from mlxtend.frequent_patterns import apriori, association_rules
from warnings import filterwarnings
filterwarnings('ignore')

In [3]:
# Read the armut_data.csv file and copy it.

df_ = pd.read_csv("armut_data.csv")
df = df_.copy()
df.head(10)

UserId  ServiceId  CategoryId           CreateDate
0   25446          4           5  2017-08-06 16:11:00
1   22948         48           5  2017-08-06 16:12:00
2   10618          0           8  2017-08-06 16:13:00
3    7256          9           4  2017-08-06 16:14:00
4   25446         48           5  2017-08-06 16:16:00
5   14354         15           1  2017-08-06 16:27:00
6   14162         21           5  2017-08-06 16:28:00
7   21230         46           4  2017-08-06 16:34:00
8   25446          6           7  2017-08-06 16:39:00
9   10659          4           5  2017-08-06 16:44:00

In [5]:
# ServiceID represents a different service for each CategoryID. \
# Create a new variable to represent services by combining ServiceID and CategoryID with “_”.
df["Hizmet"] = [str(row[1]) + "_" + str(row[2]) for row in df.values]
df.head()

UserId  ServiceId  CategoryId           CreateDate Hizmet
0   25446          4           5  2017-08-06 16:11:00    4_5
1   22948         48           5  2017-08-06 16:12:00   48_5
2   10618          0           8  2017-08-06 16:13:00    0_8
3    7256          9           4  2017-08-06 16:14:00    9_4
4   25446         48           5  2017-08-06 16:16:00   48_5

The data set consists of the date and time when the services were received, there is no basket definition (invoice etc.).In order to apply `Association Rule Learning`,\
it is necessary to create a basket (invoice etc.)definition.Here the basket definition is the services that each customer receives monthly.\
For example, 4_5, 48_5, 6_7, 47_7 services that the customer with id 25446 received in the 8th month of 2017 refers to one basket; 17_5, 14_7 services received in the 9th month of 2017 refers to another basket.\The baskets must be defined with a unique ID. For this, first create a new date variable containing only year and month. \
Combine UserID and the date variable you just created with “_” on the basis of user and assign it to a new variable named ID.

In [8]:
df["CreateDate"] = pd.to_datetime(df["CreateDate"])
df["NEW_DATE"] = df["CreateDate"].dt.strftime("%Y-%m")
df["SepetID"] = [str(row[0]) + "_" + str(row[5]) for row in df.values]
df.head()

UserId  ServiceId  CategoryId          CreateDate Hizmet NEW_DATE  \
0   25446          4           5 2017-08-06 16:11:00    4_5  2017-08   
1   22948         48           5 2017-08-06 16:12:00   48_5  2017-08   
2   10618          0           8 2017-08-06 16:13:00    0_8  2017-08   
3    7256          9           4 2017-08-06 16:14:00    9_4  2017-08   
4   25446         48           5 2017-08-06 16:16:00   48_5  2017-08   

         SepetID  
0  25446_2017-08  
1  22948_2017-08  
2  10618_2017-08  
3   7256_2017-08  
4  25446_2017-08

### Create the Association Rules Learning

In [11]:
# it must like this 

# Hizmet         0_8  10_9  11_11  12_7  13_11  14_7  15_1  16_8  17_5  18_4..
# SepetID
# 0_2017-08        0     0      0     0      0     0     0     0     0     0..
# 0_2017-09        0     0      0     0      0     0     0     0     0     0..
# 0_2018-01        0     0      0     0      0     0     0     0     0     0..
# 0_2018-04        0     0      0     0      0     1     0     0     0     0..
# 10000_2017-08    0     0      0     0      0     0     0     0     0     0..

In [12]:
invoice_product_df = df.groupby(['SepetID', 'Hizmet'])['Hizmet'].count().unstack().fillna(0).applymap(lambda x: 1 if x > 0 else 0)
invoice_product_df

Hizmet         0_8  10_9  11_11  12_7  13_11  14_7  15_1  16_8  17_5  18_4  \
SepetID                                                                      
0_2017-08        0     0      0     0      0     0     0     0     0     0   
0_2017-09        0     0      0     0      0     0     0     0     0     0   
0_2018-01        0     0      0     0      0     0     0     0     0     0   
0_2018-04        0     0      0     0      0     1     0     0     0     0   
10000_2017-08    0     0      0     0      0     0     0     0     0     0   
...            ...   ...    ...   ...    ...   ...   ...   ...   ...   ...   
99_2017-12       1     0      0     0      0     0     0     0     0     0   
99_2018-01       1     0      0     0      0     0     0     0     0     0   
99_2018-02       1     0      0     0      0     0     0     0     0     0   
9_2018-03        0     0      0     0      0     0     0     0     0     0   
9_2018-04        0     0      0     0      0     0     0     0     0     0   

Hizmet         19_6  1_4  20_5  21_5  22_0  23_10  24_10  25_0  26_7  27_7  \
SepetID                                                                      
0_2017-08         0    0     0     0     0      0      0     0     0     0   
0_2017-09         0    0     0     0     0      0      0     0     0     0   
0_2018-01         0    0     0     0     0      0      0     0     0     0   
0_2018-04         0    0     0     0     0      0      0     0     0     0   
10000_2017-08     0    0     0     0     0      0      0     0     0     0   
...             ...  ...   ...   ...   ...    ...    ...   ...   ...   ...   
99_2017-12        0    0     1     0     0      0      0     0     0     0   
99_2018-01        0    0     1     0     0      0      0     0     0     0   
99_2018-02        0    0     0     0     0      0      0     0     0     0   
9_2018-03         0    0     0     0     0      0      0     1     0     0   
9_2018-04         1    0     0     0     0      0      0     1     0     0   

Hizmet         28_4  29_0  2_0  30_2  31_6  32_4  33_4  34_6  35_11  36_1  \
SepetID                                                                     
0_2017-08         0     0    0     0     0     0     0     0      0     0   
0_2017-09         0     0    0     0     0     0     0     0      0     0   
0_2018-01         0     0    0     1     0     0     0     0      0     0   
0_2018-04         0     0    0     1     0     0     0     0      0     0   
10000_2017-08     0     0    0     0     0     0     0     0      0     0   
...             ...   ...  ...   ...   ...   ...   ...   ...    ...   ...   
99_2017-12        0     0    0     0     0     0     0     0      0     0   
99_2018-01        0     0    0     0     0     0     0     0      0     0   
99_2018-02        0     0    0     0     0     0     0     0      0     0   
9_2018-03         0     0    0     0     0     0     0     0      0     0   
9_2018-04         0     0    0     0     0     0     0     0      0     0   

Hizmet         37_0  38_4  39_10  3_5  40_8  41_3  42_1  43_2  44_0  45_6  \
SepetID                                                                     
0_2017-08         0     0      0    0     0     0     0     0     0     0   
0_2017-09         0     0      0    0     0     0     0     0     0     0   
0_2018-01         0     0      0    0     0     0     0     1     0     0   
0_2018-04         0     0      0    0     0     0     0     1     0     0   
10000_2017-08     0     0      0    0     0     0     0     0     0     0   
...             ...   ...    ...  ...   ...   ...   ...   ...   ...   ...   
99_2017-12        0     0      0    0     0     0     0     0     0     0   
99_2018-01        0     0      0    0     0     0     0     0     0     0   
99_2018-02        0     0      0    0     0     0     0     0     0     0   
9_2018-03         0     0      0    0     0     0     0     0     0     0   
9_2018-04         0     0      0    0     0 

In [15]:
frequent_itemsets = apriori(invoice_product_df, min_support=0.01, use_colnames=True)
rules = association_rules(frequent_itemsets, metric="support", min_threshold=0.01)
rules.head()

antecedents consequents  antecedent support  consequent support   support  \
0       (2_0)     (13_11)            0.130286            0.056627  0.012819   
1     (13_11)       (2_0)            0.056627            0.130286  0.012819   
2       (2_0)      (15_1)            0.130286            0.120963  0.033951   
3      (15_1)       (2_0)            0.120963            0.130286  0.033951   
4      (33_4)      (15_1)            0.027310            0.120963  0.011233   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.098394  1.737574  0.005442    1.046325       0.488074  
1    0.226382  1.737574  0.005442    1.124216       0.449965  
2    0.260588  2.154278  0.018191    1.188833       0.616073  
3    0.280673  2.154278  0.018191    1.209066       0.609539  
4    0.411311  3.400299  0.007929    1.493211       0.725728

In [17]:
# Use the arl_recommender function to recommend a service to a user who has received service 2_0 in the last 1 month.
def arl_recommender(rules_df, product_id, rec_count=1):
    sorted_rules = rules_df.sort_values("lift", ascending=False)
    recommendation_list = []
    for i, product in sorted_rules["antecedents"].items():
        for j in list(product):
            if j == product_id:
                recommendation_list.append(list(sorted_rules.iloc[i]["consequents"]))
    recommendation_list = list({item for item_list in recommendation_list for item in item_list})
    return recommendation_list[:rec_count]

arl_recommender(rules, "2_0", 1)

['2_0']

# The End